lib/system/freertos/xlnx_common/zynqmp_aarch64/sys.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2023 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * @file	freertos/xlnx_common/zynqmp_aarch64/sys.c
+ * @brief	machine specific system primitives implementation.
+ */
+
+#include <metal/compiler.h>
+#include <metal/io.h>
+#include <metal/sys.h>
+#include <metal/utilities.h>
+#include <stdint.h>
+#include "xil_cache.h"
+#include "xil_exception.h"
+#include "xil_mmu.h"
+#include "xscugic.h"
+
+/* System Device Tree (SDT) flow does not have the files generated. */
+#ifndef SDT
+
+#ifdef VERSAL_NET
+#include "xcpu_cortexa78.h"
+#elif defined(versal)
+#include "xcpu_cortexa72.h"
+#else
+#include "xreg_cortexa53.h"
+#endif /* defined(versal) */
+
+#endif /* !SDT */
+
+void sys_irq_restore_enable(unsigned int flags)
+{
+	Xil_ExceptionEnableMask(~flags);
+}
+
+unsigned int sys_irq_save_disable(void)
+{
+	unsigned int state = mfcpsr() & XIL_EXCEPTION_ALL;
+
+	if (state != XIL_EXCEPTION_ALL)
+		Xil_ExceptionDisableMask(XIL_EXCEPTION_ALL);
+
+	return state;
+}
+
+void metal_machine_cache_flush(void *addr, unsigned int len)
+{
+	if (!addr || !len)
+		Xil_DCacheFlush();
+	else
+		Xil_DCacheFlushRange((intptr_t)addr, len);
```

Comment:
Here too for cache flush. We can sync with developers who are working on Xilinx's freertos port to ask correct API that should be used to perform all operations in this file.

Reply:
this part I am would push back for 2 reasons:
1. This API is already in libmetal for other ports. 
2. There is no alternative for these APIs in Xilinx FreeRTOS port.

Reply:
That is what I am trying to say. We need to introduce cache support in FreeRTOS and then use those API here. So, metal_machine_cache_flush interface will remain same, but implementation should be freertos specific.

Reply:
If FreeRTOS call isn't available, then that API doesn't belong in this directory. This directory is supposed to translate libmetal calls to freertos calls. Can't introduce baremtal calls to platforms here. Please move such APIs under generic/<platform>/directory.

Reply:
For versal-net support, we need generic/zynqmp_r52 directory and have calls there.

Reply:
Ok I checked FreeRTOS code, they don't provide any standard FreeRTOS specific framework for Cache Management instead, it is upto platforms how to implement it. So, in that case we have to use Xilinx BSP APIs here.